### DISCUSSION TITLE
How can I set the font for individual candidates/characters?

### DISCUSSION THREAD

**Discussion by Rohupt** [saschanaz/ime-rs#10] [General]
I have a large dictionary that has unique glyphs that no font can include them all. I'm aware of how to set the default font for the program, but can I set a list of fonts for fallback - that is, if a certain glyph in a candidate doesn't present in a font, the program will fall back to the next one in the list for that particular glyph/candidate?

Still getting the hang of Rust here, please be easy on me.

**Comment by saschanaz** [comment]
I don't think the current setup is advanced enough for that scenario. Per [the Microsoft documentation](https://docs.microsoft.com/en-us/windows/win32/intl/using-font-fallback) the app should call `ScriptShape()` to check whether the given font supports the text or not, but the demo app never calls that.

`CCandidateWindow::_OnPaint` can probably be tweaked, which unfortunately hasn't been ported to Rust yet.

**Comment by Rohupt** [comment]
Well, [this here](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-exttextoutw) says "ExtTextOut will use [Uniscribe](https://docs.microsoft.com/en-us/windows/win32/intl/uniscribe) when necessary resulting in font fallback. The ETO_IGNORELANGUAGE flag will inhibit this behavior and should not be passed.", means the fallback function should have been already implemented because ExtTextOut is used to write the candidate list into the candidate window. And yet it doesn't show the unsupported characters (just blank) when I run the IME. Do you have any comment about this?

**Reply by saschanaz** [reply]
I have no idea then. Do you have an example text and the expected font fallback behavior (including each font name)?

**Reply by Rohupt** [reply]
[NomNaTong-Regular.zip](https://github.com/saschanaz/ime-rs/files/9274457/NomNaTong-Regular.zip)
I'm trying to add some Greek characters to the dictionary, which, of course, are not supported by this Vietnamese Nom font. `∑σς` for example.

Someone reported [here](https://webcache.googleusercontent.com/search?q=cache:nf8M7rtPlMcJ:https://groups.google.com/g/microsoft.public.win32.programmer.gdi/c/szhVjUab9U0+) that "For example, draw a mixture of English, Japanese, and Arabic text (in one Unicode string) with ExtTextOut, using Arial. Japanese Glyphs are not present in the Arial, but they are shown anyway." I understand this as that I don't have to specify fallback fonts, Windows will automatically choose them and replace unavailable glyphs. I tried to convert the font to OTF as stated in the link as well, but either case (TTF or OTF) it doesn't work.

For example, with the value `σsσs`, the result is like this:

![image](https://user-images.githubusercontent.com/34348907/183236420-538158b1-987e-4f55-b466-68c15e426e15.png)

I can't interfere in whatever `ExtTextOut` does to change the font mid-string, so maybe I'll have to rewrite it as a whole.
